File: sdk/src/test/java/com/webimapp/android/sdk/impl/MessageHolderImplTest.java
package com.webimapp.android.sdk.impl;

import com.webimapp.android.sdk.Message;
import com.webimapp.android.sdk.MessageListener;
import com.webimapp.android.sdk.MessageTracker;
import com.webimapp.android.sdk.impl.items.ChatItem;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class MessageHolderImplTest {

    private Runnable EMPTY_RUNNABLE = () -> {};

    private List<MessageImpl> EMPTY_MESSAGES_LIST = Collections.emptyList();

    private <T> Set<T> setOf(T ... elements) {
        return new HashSet<>(Arrays.asList(elements));
    }

    private <T> List<T> listOf(T ... elements) {
        return Arrays.asList(elements);
    }

    private ArgumentCaptor<List<Message>> newListCaptor() {
        return ArgumentCaptor.forClass(List.class);
    }

    private ArgumentCaptor<Message> newMessageCaptor() {
        return ArgumentCaptor.forClass(Message.class);
    }

    static class History implements RemoteHistoryProvider {
        List<MessageImpl> history;

        History(List<MessageImpl> history) {
            this.history = history;
        }

        @Override
        public void requestHistoryBefore(long beforeMessageTs, HistoryBeforeCallback callback) {
            int before = -1;
            for (int i = history.size() - 1; i >= 0; i--) {
                MessageImpl historyMessage = history.get(i);
                if (historyMessage.getTimeMicros() == beforeMessageTs) {
                    before = i;
                    break;
                } else if (historyMessage.getTimeMicros() < beforeMessageTs) {
                    before = i + 1;
                    break;
                }
            }
            int after = Math.max(0, before - 100);
            callback.onSuccess(
                before <= 0
                    ? Collections.emptyList()
                    : history.subList(after, before),
                after != 0
            );
        }
    }

    private List<MessageImpl> generateHistoryInRange(int first, int last) {
        List<MessageImpl> history = new ArrayList<>();
        for (int i = first; i < last; i++) {
            history.add(new MessageImpl(
                "",
                StringId.forMessage(String.valueOf(i)),
                null,
                StringId.forOperator("op"),
                "",
                "",
                Message.Type.OPERATOR,
                "text",
                i,
                String.valueOf(i),
                null,
                true,
                null,
                false,
                false,
                false,
                false,
                null,
                null,
                null,
                null
            ));
        }
        return history;
    }

    private List<MessageImpl> generateCurrentChatInRange(int first, int last, Set<Integer> expect) {
        List<MessageImpl> currentChatMessages = new ArrayList<>();
        for (int i = first; i < last; i++) {
            if (expect.contains(i)) {
                continue;
            }
            currentChatMessages.add(new MessageImpl(
                "",
                StringId.forMessage(String.valueOf(i)),
                null,
                StringId.forOperator("op"),
                "",
                "",
                Message.Type.OPERATOR,
                "text",
                i,
                String.valueOf(i),
                null,
                false,
                null,
                false,
                true,
                false,
                false,
                null,
                null,
                null,
                null
            ));
        }
        return currentChatMessages;
    }

    private List<MessageImpl> generateHistoryFromCurrentChat(List<? extends MessageImpl> input) {
        List<MessageImpl> historyMessages = new ArrayList<>();
        for (MessageImpl currentChatMessage : input) {
            historyMessages.add(
                new MessageImpl(
                    "",
                    currentChatMessage.getClientSideId(),
                    currentChatMessage.getSessionId(),
                    currentChatMessage.getOperatorId(),
                    currentChatMessage.getAvatarUrlLastPart(),
                    currentChatMessage.getSenderName(),
                    currentChatMessage.getType(),
                    currentChatMessage.getText(),
                    currentChatMessage.timeMicros,
                    String.valueOf(currentChatMessage.timeMicros),
                    null,
                    true,
                    null,
                    false,
                    true,
                    false,
                    false,
                    null,
                    null,
                    null,
                    null
                )
            );
        }
        return historyMessages;
    }

    private MessageImpl newCurrentChatMessage(int index) {
        return new MessageImpl(
            "",
            StringId.forMessage(String.valueOf(index)),
            null,
            StringId.forOperator("op"),
            "",
            "",
            Message.Type.OPERATOR,
            "text",
            index,
            String.valueOf(index),
            null,
            false,
            null,
            false,
            true,
            false,
            false,
            null,
            null,
            null,
            null
        );
    }

    private MessageImpl newEditedCurrentChatMessage(MessageImpl original) {
        return new MessageImpl(
            "",
            original.getClientSideId(),
            original.getSessionId(),
            original.getOperatorId(),
            original.getAvatarUrlLastPart(),
            original.getSenderName(),
            original.getType(),
            original.getText() + "1",
            original.getTimeMicros(),
            original.getServerSideId(),
            null,
            false,
            null,
            false,
            true,
            false,
            false,
            null,
            null,
            null,
            null
        );
    }

    private MessageImpl newEditedHistoryMessage(MessageImpl original) {
        return new MessageImpl(
            "",
            original.getClientSideId(),
            original.getSessionId(),
            original.getOperatorId(),
            original.getAvatarUrlLastPart(),
            original.getSenderName(),
            original.getType(),
            original.getText() + "1",
            original.getTimeMicros(),
            original.getServerSideId(),
            null,
            true,
            null,
            false,
            true,
            false,
            false,
            null,
            null,
            null,
            null
        );
    }

    private MessageHolderImpl newMessageHolderWithHistory(RemoteHistoryProvider history) {
        return new MessageHolderImpl(AccessChecker.EMPTY, history, new MemoryHistoryStorage(), false);
    }

    private MessageHolderImpl newMessageHolderWithHistory(RemoteHistoryProvider history, List<MessageImpl> localHistory) {
        return new MessageHolderImpl(AccessChecker.EMPTY, history, new MemoryHistoryStorage(localHistory), false);
    }

    private List<MessageImpl> concat(List<MessageImpl> list1, List<MessageImpl> list2) {
        List<MessageImpl> resultList = new ArrayList<>();
        resultList.addAll(list1);
        resultList.addAll(list2);
        return resultList;
    }

    private List<MessageImpl> cloneCurrentChatMessages(List<MessageImpl> currentMessages) {
        try {
            Method method = currentMessages.getClass().getDeclaredMethod("clone");
            method.setAccessible(true);
            return (List<MessageImpl>) method.invoke(currentMessages);
        } catch (Exception exception) {
            exception.printStackTrace();
            return null;
        }
    }

    private void setPrivateField(String privateFieldName, Object privateFieldValue, Object object) {
        try {
            Field field = object.getClass().getDeclaredField(privateFieldName);
            field.setAccessible(true);
            field.set(object, privateFieldValue);
        } catch (Exception exception) {
            exception.printStackTrace();
        }
    }

    @Test
    public void internalHistoryGeneration() {
        List<MessageImpl> history1 = generateHistoryInRange(0, 10);
        List<MessageImpl> history3 = generateHistoryInRange(20, 30);

        List<String> ids1 = history1.stream()
            .map(MessageImpl::getServerSideId)
            .collect(Collectors.toList());

        List<String> ids3 = history3.stream()
            .map(MessageImpl::getServerSideId)
            .collect(Collectors.toList());

        assertEquals(ids1, IntStream.rangeClosed(0, 9).boxed().map(String::valueOf).collect(Collectors.toList()));
        assertEquals(ids3, IntStream.rangeClosed(20, 29).boxed().map(String::valueOf).collect(Collectors.toList()));
    }

    @Test
    public void respondCurrentChatMessagesImmediately() {
        List<MessageImpl> currentChatMessages = generateCurrentChatInRange(0, 10, setOf());
        MessageHolder messageHolder = newMessageHolderWithHistory(new History(Collections.<MessageImpl>emptyList()));
        MessageListener messageListener = mock(MessageListener.class);
        MessageTracker messageTracker = messageHolder.newMessageTracker(messageListener);
        MessageTracker.GetMessagesCallback callback = mock(MessageTracker.GetMessagesCallback.class);

        messageHolder.onChatReceive(null, new ChatItem(), currentChatMessages);

        messageTracker.getNextMessages(10, callback);
        verify(callback, times(1)).receive(currentChatMessages);
    }

    @Test
    public void checkGetNextMessages() {
        List<MessageImpl> currentChatMessages = generateCurrentChatInRange(0, 10, setOf());
        List<MessageImpl> history = generateHistoryFromCurrentChat(currentChatMessages);
        MessageHolder messageHolder = newMessageHolderWithHistory(new History(listOf()));
        MessageListener messageListener = mock(MessageListener.class);
        MessageTracker messageTracker = messageHolder.newMessageTracker(messageListener);
        MessageTracker.GetMessagesCallback callback = mock(MessageTracker.GetMessagesCallback.class);

        // This method caches callback
        messageTracker.getNextMessages(10, callback);
        // Callback will be cached because local storage is empty.
        verify(callback, never()).receive(Mockito.any());

        reset(callback);
        messageHolder.onChatReceive(null, new ChatItem(), currentChatMessages);
        // Our cached callback will be fired.
        verify(callback, times(1)).receive(currentChatMessages);

        reset(messageListener);
        // History messages will be saved to local storage and merged with currentMessages
        messageHolder.receiveHistoryUpdate(history, setOf(), EMPTY_RUNNABLE);
        verify(messageListener, times(currentChatMessages.size())).messageChanged(Mockito.any(), Mockito.any());
    }

    @Test
    public void checkGetAllMessages() {
        List<MessageImpl> history1 = generateHistoryInRange(0, 10);
        List<MessageImpl> currentChatMessages = generateCurrentChatInRange(10, 20, setOf());
        List<MessageImpl> history2 = generateHistoryFromCurrentChat(currentChatMessages);
        MessageHolder messageHolder = newMessageHolderWithHistory(new History(listOf()));
        MessageListener messageListener = mock(MessageListener.class);
        MessageTracker messageTracker = messageHolder.newMessageTracker(messageListener);
        MessageTracker.GetMessagesCallback callback = mock(MessageTracker.GetMessagesCallback.class);

        // This method doesn't cache callback, it just gets all messages from local storage
        messageTracker.getAllMessages(callback);
        // Local storage is empty, so we get empty list
        verify(callback, times(1)).receive(EMPTY_MESSAGES_LIST);

        messageHolder.onChatReceive(null, new ChatItem(), currentChatMessages);
        // history1 will be saved to local storage
        messageHolder.receiveHistoryUpdate(history1, setOf(), EMPTY_RUNNABLE);
        // history2 will be saved to local storage
        messageHolder.receiveHistoryUpdate(history2, setOf(), EMPTY_RUNNABLE);

        reset(callback);
        messageTracker.getAllMessages(callback);
        verify(callback, times(1)).receive(concat(history1, history2));
    }

    @Test
    public void checkGetLastMessages() {
        List<MessageImpl> currentChatMessages = generateCurrentChatInRange(0, 10, setOf());
        List<MessageImpl> history1 = generateHistoryFromCurrentChat(currentChatMessages);
        List<MessageImpl> history2 = generateHistoryInRange(10, 20);
        MessageHolder messageHolder = newMessageHolderWithHistory(new History(Collections.<MessageImpl>emptyList()));
        MessageListener messageListener = mock(MessageListener.class);
        MessageTracker messageTracker = messageHolder.newMessageTracker(messageListener);
        MessageTracker.GetMessagesCallback callback = mock(MessageTracker.GetMessagesCallback.class);

        // This method caches callback
        messageTracker.getLastMessages(10, callback);
        // Local storage is empty, so callback is cached and will not be fired
        verify(callback, never()).receive(Mockito.any());

        reset(callback);
        messageHolder.onChatReceive(null, new ChatItem(), currentChatMessages);
        // Fire cached callback
        verify(callback, times(1)).receive(currentChatMessages);

        reset(messageListener);
        messageHolder.receiveHistoryUpdate(history1, setOf(), EMPTY_RUNNABLE);
        verify(messageListener, times(history1.size())).messageChanged(Mockito.any(), Mockito.any());
        messageHolder.receiveHistoryUpdate(history2, setOf(), EMPTY_RUNNABLE);
        verify(messageListener, times(history2.size())).messageAdded(Mockito.isNull(), Mockito.any());

        reset(callback);
        messageTracker.getLastMessages(10, callback);
        verify(callback, times(1)).receive(history2);
    }

    @Test
    public void checkResetTo() {
        List<MessageImpl> currentChatMessages = generateCurrentChatInRange(0, 10, setOf());
        MessageHolder messageHolder = newMessageHolderWithHistory(new History(Collections.<MessageImpl>emptyList()));
        MessageListener messageListener = mock(MessageListener.class);
        MessageTracker messageTracker = messageHolder.newMessageTracker(messageListener);
        MessageTracker.GetMessagesCallback callback = mock(MessageTracker.GetMessagesCallback.class);

        messageHolder.onChatReceive(null, new ChatItem(), currentChatMessages);

        messageTracker.getNextMessages(8, callback);
        verify(callback, times(1)).receive(currentChatMessages.subList(2, 10));

        // Now headMessage is set to message with index '2', let's reset currentMessages to '5'
        reset(messageListener);
        messageTracker.resetTo(currentChatMessages.get(5));
        // Check that for messages from '2' to '5' was called callback
        verify(messageListener, times(3)).messageRemoved(any());

        reset(callback);
        messageTracker.getNextMessages(5, callback);
        // Here we must get messages from '5' to '0'
        verify(callback, times(1)).receive(currentChatMessages.subList(0, 5));

        reset(messageListener);
        messageTracker.resetTo(currentChatMessages.get(9));
        verify(messageListener, times(9)).messageRemoved(any());

        reset(callback);
        messageTracker.getNextMessages(10, callback);
        verify(callback, times(1)).receive(currentChatMessages.subList(0, 9));
    }

    @Test
    public void checkMergeCurrentChatWith() {
        List<MessageImpl> currentChatMessages1 = generateCurrentChatInRange(0,10, setOf(1, 2, 3));
        List<MessageImpl> currentChatMessages2 = generateCurrentChatInRange(0,10, setOf(8));
        MessageHolder messageHolder = newMessageHolderWithHistory(new History(Collections.<MessageImpl>emptyList()));
        MessageListener messageListener = mock(MessageListener.class);
        MessageTracker messageTracker = messageHolder.newMessageTracker(messageListener);
        MessageTracker.GetMessagesCallback callback = mock(MessageTracker.GetMessagesCallback.class);

        ChatItem chatItem = new ChatItem();
        // set these private fields, to make chat items equals
        setPrivateField("id", "1id", chatItem);
        setPrivateField("clientSideId", "1csi", chatItem);

        assertEquals(chatItem, chatItem);

        messageHolder.onChatReceive(null, chatItem, currentChatMessages1);
        messageTracker.getNextMessages(10, callback);
        verify(callback, times(1)).receive(currentChatMessages1);

        reset(messageListener);
        // Here we merging to equals chats with different currentChatMessages
        messageHolder.onChatReceive(chatItem, chatItem, currentChatMessages2);
        verify(messageListener, times(3)).messageAdded(any(), any()); // for 1, 2, 3
        verify(messageListener, times(1)).messageRemoved(any()); // for 8
    }

    @Test
    public void checkMessageAddedCallbackFired() {
        List<MessageImpl> currentMessages = generateCurrentChatInRange(0,10, setOf(8));
        MessageHolder messageHolder = newMessageHolderWithHistory(new History(Collections.<MessageImpl>emptyList()));
        MessageListener messageListener = mock(MessageListener.class);
        MessageTracker messageTracker = messageHolder.newMessageTracker(messageListener);
        MessageTracker.GetMessagesCallback callback = mock(MessageTracker.GetMessagesCallback.class);

        messageHolder.onChatReceive(null, new ChatItem(), currentMessages);
        // call this method to set MessageTracker#headMessage pointer first current message '0',
        // because MessageListener fires callbacks only if message placed before headMessage
        messageTracker.getNextMessages(10, callback);

        reset(messageListener);
        MessageImpl newMessage = newCurrentChatMessage(11);
        // Add message to end of chat
        messageHolder.onMessageAdded(newMessage);
        // Verify message added to end
        verify(messageListener, times(1)).messageAdded(null, newMessage);

        reset(messageListener);
        newMessage = newCurrentChatMessage(8); // message with id '8'
        // Add message to middle of chat
        messageHolder.onMessageAdded(newMessage);
        // Verify message added to middle
        MessageImpl beforeMessage = currentMessages.get(8); // message with id '9'
        verify(messageListener, times(1)).messageAdded(beforeMessage, newMessage);
    }


    @Test
    public void checkMessageChangedCallbackFired() {
        List<MessageImpl> currentMessages = generateCurrentChatInRange(0,10, setOf());
        MessageHolder messageHolder = newMessageHolderWithHistory(new History(Collections.<MessageImpl>emptyList()));
        MessageListener messageListener = mock(MessageListener.class);
        MessageTracker messageTracker = messageHolder.newMessageTracker(messageListener);
        MessageTracker.GetMessagesCallback callback = mock(MessageTracker.GetMessagesCallback.class);

        messageHolder.onChatReceive(null, new ChatItem(), currentMessages);
        // call this method to set MessageTracker#headMessage pointer first current message '0',
        // because MessageListener fires callbacks only if message placed before headMessage
        messageTracker.getNextMessages(10, callback);

        reset(messageListener);
        MessageImpl originalMessage = currentMessages.get(8);
        MessageImpl editedMessage = newEditedHistoryMessage(originalMessage);
        // Add message to middle of chat
        messageHolder.onMessageChanged(editedMessage);
        // Verify message was changed at position '8'
        verify(messageListener, times(1)).messageChanged(originalMessage, editedMessage);
    }


    @Test
    public void checkMessageDeletedCallbackFired() {
        List<MessageImpl> currentMessages = generateCurrentChatInRange(0,10, setOf());
        MessageHolder messageHolder = newMessageHolderWithHistory(new History(Collections.<MessageImpl>emptyList()));
        MessageListener messageListener = mock(MessageListener.class);
        MessageTracker messageTracker = messageHolder.newMessageTracker(messageListener);
        MessageTracker.GetMessagesCallback callback = mock(MessageTracker.GetMessagesCallback.class);

        messageHolder.onChatReceive(null, new ChatItem(), currentMessages);
        // call this method to set MessageTracker#headMessage pointer first current message '0',
        // because MessageListener fires callbacks only if message placed before headMessage
        messageTracker.getNextMessages(10, callback);

        reset(messageListener);
        MessageImpl messageToDelete = currentMessages.get(5);
        // Delete message from middle of chat
        messageHolder.onMessageDeleted(messageToDelete.getServerSideId());
        // Verify message deleted from middle
        verify(messageListener, times(1)).messageRemoved(messageToDelete);
    }
}

File: sdk/src/test/java/ru/webim/android/sdk/impl/MessageHolderImplTest.kt
package ru.webim.android.sdk.impl

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner
import org.mockito.kotlin.*
import ru.webim.android.sdk.Message
import ru.webim.android.sdk.MessageListener
import ru.webim.android.sdk.MessageTracker
import ru.webim.android.sdk.MessageTracker.GetMessagesCallback
import ru.webim.android.sdk.impl.items.ChatItem
import kotlin.math.max

@RunWith(MockitoJUnitRunner::class)
class MessageHolderImplTest {
    private val messageCaptor: KArgumentCaptor<Message>
        get() = argumentCaptor()
    private val listCaptor: KArgumentCaptor<List<Message>>
        get() = argumentCaptor()

    fun mockers(messageHolder: MessageHolder): Triple<MessageTracker, MessageListener, GetMessagesCallback> {
        val listener = Mockito.mock(MessageListener::class.java)
        val callback = Mockito.mock(GetMessagesCallback::class.java)
        val tracker = messageHolder.newMessageTracker(listener)
        return Triple(tracker, listener, callback)
    }

    @Test
    fun checkHistoryGeneration() {
        val ids1 = generateChat(0, 10, isHistory = true).map { it.serverSideId.toInt() }
        val ids3 = generateChat(20, 30, isHistory = true).map { it.serverSideId.toInt() }

        assertEquals(ids1, (0..9).toList())
        assertEquals(ids3, (20..29).toList())
    }

    @Test
    fun respondCurrentChatMessagesImmediately() {
        val currentMsgs = generateChat(0, 10)
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        holder.onChatReceive(null, ChatItem(), currentMsgs)
        tracker.getNextMessages(10, callback)

        verify(callback, times(1)).receive(currentMsgs)
    }

    @Test
    fun checkGetNextMessages() {
        val currentMsgs = generateChat(0, 10)
        val historyMsgs = generateChat(0, 10, isHistory = true)
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        // This method caches callback
        tracker.getNextMessages(10, callback)
        // Callback will be cached because local storage is empty.
        verify(callback, never()).receive(any())
        reset(callback)

        // Our cached callback will be fired.
        holder.onChatReceive(null, ChatItem(), currentMsgs)
        verify(callback, times(1)).receive(currentMsgs)
        reset(listener)

        // History messages will be saved to local storage and merged with currentMessages
        holder.receiveHistoryUpdate(historyMsgs, setOf()) {}
        verify(listener, times(currentMsgs.size)).messageChanged(any(), any())
    }

    @Test
    fun checkGetAllMessages() {
        val history1 = generateChat(0, 10, isHistory = true)
        val history2 = generateChat(10, 20, isHistory = true)
        val currentMsgs = generateChat(10, 20)
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        // This method doesn't cache callback, it just gets all messages from local storage
        tracker.getAllMessages(callback)
        // Local storage is empty, so we get empty list
        verify(callback, times(1)).receive(emptyList())
        holder.onChatReceive(null, ChatItem(), currentMsgs)
        // history1 will be saved to local storage
        holder.receiveHistoryUpdate(history1, setOf()) {}
        // history2 will be saved to local storage
        holder.receiveHistoryUpdate(history2, setOf()) {}
        reset(callback)

        tracker.getAllMessages(callback)
        verify(callback, times(1)).receive(history1 + history2)
    }

    @Test
    fun checkGetLastMessages() {
        val currentChatMessages = generateChat(0, 10)
        val history1 = generateChat(0, 10, isHistory = true)
        val history2 = generateChat(10, 20)
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        // This method caches callback
        tracker.getLastMessages(10, callback)
        // Local storage is empty, so callback is cached and will not be fired
        verify(callback, never()).receive(any())
        reset(callback)

        holder.onChatReceive(null, ChatItem(), currentChatMessages)
        // Fire cached callback
        verify(callback, times(1)).receive(currentChatMessages)
        reset(listener)

        holder.receiveHistoryUpdate(history1, setOf()) {}
        verify(listener, times(history1.size)).messageChanged(any(), any())
        holder.receiveHistoryUpdate(history2, setOf()) {}
        verify(listener, times(history2.size)).messageAdded(isNull(), any())
        reset(callback)

        tracker.getLastMessages(10, callback)
        verify(callback, times(1)).receive(history2)
    }

    @Test
    fun checkResetTo() {
        val currentMsgs = generateChat(0, 10)
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        holder.onChatReceive(null, ChatItem(), currentMsgs)
        tracker.getNextMessages(8, callback)
        verify(callback, times(1)).receive(currentMsgs.subList(2, 10))

        // Now headMessage is set to message with index '2', let's reset currentMessages to '5'
        reset(listener)
        tracker.resetTo(currentMsgs[5])
        // Check that for messages from '2' to '5' was called callback
        verify(listener, times(3)).messageRemoved(any())
        reset(callback)

        tracker.getNextMessages(5, callback)
        // Here we must get messages from '5' to '0'
        verify(callback, times(1)).receive(currentMsgs.subList(0, 5))
        reset(listener)

        tracker.resetTo(currentMsgs[9])
        verify(listener, times(9)).messageRemoved(any())
        reset(callback)

        tracker.getNextMessages(10, callback)
        verify(callback, times(1)).receive(currentMsgs.subList(0, 9))
    }

    @Test
    fun checkMergeCurrentChatWith() {
        val current1 = generateChat(0, 10, setOf(1, 2, 3))
        val current2 = generateChat(0, 10, setOf(8))
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        // set these private fields, to make chat items equals
        val chatItem = ChatItem().apply {
            setPrivateField("id", "1id")
            setPrivateField("clientSideId", "1csi")
        }
        assertEquals(chatItem, chatItem)

        holder.onChatReceive(null, chatItem, current1)
        tracker.getNextMessages(10, callback)
        verify(callback, times(1)).receive(current1)
        reset(listener)
        // Here we merging to equals chats with different currentChatMessages
        holder.onChatReceive(chatItem, chatItem, current2)
        // for 1, 2, 3
        verify(listener, times(3)).messageAdded(any(), any())
        // for 8
        verify(listener, times(1)).messageRemoved(any())
    }

    @Test
    fun checkMessageAddedCallbackFired() {
        val currentMessages = generateChat(0, 10, setOf(8))
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        holder.onChatReceive(null, ChatItem(), currentMessages)
        // call this method to set MessageTracker#headMessage pointer first current message '0',
        // because MessageListener fires callbacks only if message placed before headMessage
        tracker.getNextMessages(10, callback)
        reset(listener)
        var newMessage: MessageImpl? = message(11)
        // Add message to end of chat
        holder.onMessageAdded(newMessage!!)
        // Verify message added to end
        verify(listener, times(1)).messageAdded(null, newMessage)
        reset(listener)
        newMessage = message(8) // message with id '8'
        // Add message to middle of chat
        holder.onMessageAdded(newMessage)
        // Verify message added to middle
        val beforeMessage = currentMessages[8] // message with id '9'
        verify(listener, times(1)).messageAdded(beforeMessage, newMessage)
    }

    @Test
    fun checkMessageChangedCallbackFired() {
        val currentMsgs = generateChat(0, 10)
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        holder.onChatReceive(null, ChatItem(), currentMsgs)
        // call this method to set MessageTracker#headMessage pointer first current message '0',
        // because MessageListener fires callbacks only if message placed before headMessage
        tracker.getNextMessages(10, callback)
        verify(callback).receive(currentMsgs)
        reset(callback)

        val originalMessage = currentMsgs[8]
        val editedMessage: MessageImpl = message(8, isEdited = true)
        // Add message to middle of chat
        holder.onMessageChanged(editedMessage)
        // Verify message was changed at position '8'
        verify(listener).messageChanged(originalMessage, editedMessage)
    }

    @Test
    fun checkMessageDeletedCallbackFired() {
        val currentMsgs = generateChat(0, 10)
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        holder.onChatReceive(null, ChatItem(), currentMsgs)
        // call this method to set MessageTracker#headMessage pointer first current message '0',
        // because MessageListener fires callbacks only if message placed before headMessage
        tracker.getNextMessages(10, callback)
        reset(listener)
        val messageToDelete = currentMsgs[5]
        // Delete message from middle of chat
        holder.onMessageDeleted(messageToDelete.serverSideId)
        // Verify message deleted from middle
        verify(listener, times(1)).messageRemoved(messageToDelete)
    }

    @Test
    fun checkMessagesAdded() {
        val currentMsgs = generateChat(0, 10)
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        tracker.getNextMessages(10, callback)
        verify(callback, never()).receive(any())
        reset(callback)

        holder.onChatReceive(null, null, emptyList())
        verify(listener, never()).messageAdded(any(), any())
        verify(callback).receive(emptyList())
        reset(listener)
        reset(callback)

        holder.onChatReceive(null, ChatItem(), currentMsgs)
        verify(listener, times(10)).messageAdded(anyOrNull(), any())
        verify(callback, never()).receive(any())
    }

    @Test
    fun checkListenerIsCalledForHistoryUpdatedMessages() {
        val localOldHistoryMessages = generateChat(0, 10)
        val remoteUpdatedHistoryMessages = generateChat(0, 10)
        val editedList: MutableList<MessageImpl> = mutableListOf()
        editedList.add(message(0, isEdited = true))
        editedList.add(message(1, isEdited = true))
        editedList.add(message(2, isEdited = true))
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        tracker.getNextMessages(10, callback)
        holder.onChatReceive(null, ChatItem(), localOldHistoryMessages)
        reset(listener)
        reset(callback)

        holder.receiveHistoryUpdate(editedList, setOf()) {}
        val captor = messageCaptor
        verify(listener, times(3)).messageChanged(any(), captor.capture())
        assertArrayEquals(editedList.toTypedArray(), captor.allValues.toTypedArray())
    }

    @Test
    fun checkMirrorsMessages() {
        val holder = messageHolder()
        val (tracker, listener, callback) = mockers(holder)

        holder.onChatReceive(null, null, emptyList())
        val sending1 = sendingMessage(1)
        val sending2 = sendingMessage(2)
        val sending4 = sendingMessage(4)

        holder.onSendingMessage(sending1)
        holder.onSendingMessage(sending2)
        holder.onSendingMessage(sending4)

        val message1 = message(1)
        val message2 = message(2)
        val message3 = message(3)
        val message4 = message(4)

        holder.onMessageAdded(message1)
        verify(listener).messageChanged(sending1, message1)
        reset(listener)

        holder.onMessageAdded(message4)
        verify(listener).messageChanged(sending4, message4)
        reset(listener)

        holder.onMessageAdded(message2)
        verify(listener).messageChanged(sending2, message2)
        reset(listener)

        holder.onMessageAdded(message3)
        verify(listener).messageAdded(message4, message3)
    }

    private fun generateChat(from: Int, to: Int, expect: Set<Int> = setOf(), isHistory: Boolean = false): MutableList<MessageImpl> {
        val messages: MutableList<MessageImpl> = mutableListOf()
        for (index in from until to) {
            if (expect.contains(index)) continue
            messages.add(message(index, isHistory))
        }
        return messages
    }

    private fun message(index: Int, isEdited: Boolean = false, isHistory: Boolean = false) = MessageImpl(
        "",
        StringId.forMessage(index.toString()),
        null,
        StringId.forOperator("op"),
        "",
        "",
        Message.Type.OPERATOR,
        if (!isEdited) "text $index" else "edited $index",
        index * 1000L,
        index.toString(),
        null,
        isHistory,
        null,
        false,
        true,
        false,
        isEdited,
        null,
        null,
        null,
        null,
        null,
        false,
        false
    )

    private fun sendingMessage(index: Int) = MessageSending(
        "",
        StringId.forMessage(index.toString()),
        "",
        Message.Type.VISITOR,
        index.toString(),
        index * 1000L,
        null,
        null,
        null
    )

    private fun messageHolder(serverHistory: List<MessageImpl> = listOf(), localHistory: List<MessageImpl> = listOf()): MessageHolderImpl {
        return MessageHolderImpl(AccessChecker.EMPTY, History(serverHistory), MemoryHistoryStorage(localHistory), false)
    }

    private fun Any.setPrivateField(privateFieldName: String, privateFieldValue: Any) {
        try {
            val field = javaClass.getDeclaredField(privateFieldName)
            field.isAccessible = true
            field[this] = privateFieldValue
        } catch (exception: Exception) {
            exception.printStackTrace()
        }
    }

    internal class History(var history: List<MessageImpl>) :
        RemoteHistoryProvider {
        override fun requestHistoryBefore(beforeMessageTs: Long, callback: HistoryBeforeCallback) {
            var before = -1
            for (i in history.indices.reversed()) {
                val historyMessage = history[i]
                if (historyMessage.getTimeMicros() == beforeMessageTs) {
                    before = i
                    break
                } else if (historyMessage.getTimeMicros() < beforeMessageTs) {
                    before = i + 1
                    break
                }
            }
            val after = max(0, before - 100)
            callback.onSuccess(
                if (before <= 0) emptyList() else history.subList(after, before),
                after != 0
            )
        }
    }
}